### PR TITLE
fix(router): eagerly update internal state on browser-triggered navigations

### LINKED ISSUE #43101: Returning UrlTree from CanActivate guard breaks browser's history and leads to an inconsistent state
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

No

### Description

Returning a `UrlTree` from `CanActivate` guard should return in cancelling current navigation and redirecting to a new one.

The following events result in an inconsistent state:

* User navigates through various routes
* He performs an action, old routes won't be accessible because a `CanActivate` guard now returns a `UrlTree` to a valid route
* User goes back with the browser. First time it works well and redirection to new route happens. Second time onwards the view is correct, but now the url is out of sync with current activated route .

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-router-issue?file=src/app/app.component.ts

### Please provide the exception or error you saw

```true
router.url !== window.location.pathname
```

### Please provide the environment you discovered this bug in

```true
Happens at least in v11 and in v12, I've tested with both environments
```

### Anything else?

Demo of the issue:

routing-issue

### PR DESCRIPTION
**Pull request by atscott**
The management of `browserUrlTree` currently has several problems with
correctly tracking the actual state of the browser.

This change makes the Router eagerly update the `browserUrlTree` when
handling navigations triggered by browser events (i.e., not 'imperative'). This
is because with those types of navigations, the browser URL bar is
_already_ updated. If we do not update the internal tracking of the
`browserUrlTree`, we will be out of sync with the real URL if the
navigation is rejected.

It would be best if we could remove `browserUrlTree` completely, but doing that
would require a lot more investigation and is blocked by #27059 because
the SpyLocation used in tests does not emulate real browser behavior.

fixes #43101

### CODE REVIEW

**Comment by atscott** [comment]
[global presubmit started](https://fusion2.corp.google.com/presubmit/tap/389917388/OCL:389917388:BASE:389877317:1628616937765:262ea5c8/related)

**Comment by atscott** [comment]
[global presubmit](https://test.corp.google.com/ui#id=OCL:389917388:BASE:390019790:1628642715226:2f40953e) with reruns shows no failures due to this change.

File: packages/router/src/router.ts [hunk 1]
```diff
@@ -636,6 +636,12 @@ export class Router {
                            (this.onSameUrlNavigation === 'reload' ? true : urlTransition) &&
                            this.urlHandlingStrategy.shouldProcessUrl(t.rawUrl);
 
+                       // If the source of the navigation is from a browser event, the URL is
+                       // already updated. We already need to sync the internal state.
+                       if (t.source !== 'imperative') {
```
Comment (AndrewKushnir) [inline packages/router/src/router.ts:641]: Thanks for adding comments, they help a lot with review 👍 

Small proposal (feel free to ignore): I was thinking that we can have a small helper function to make the code more self-documented:
```suggestion
                       if (isBrowserTriggeredNavigation(t.source)) {
```
Such function can be used in a couple places in this PR and also in other parts of the codebase.

**Review by AndrewKushnir — APPROVED** [review]
👍

**Comment by atscott** [comment]
merge assistance: while global presubmit ran and was green, this CL should still be merged and synced on its own to be safe.

**Comment by AndrewKushnir** [comment]
@atscott FYI there are some CI failures which look legit, could you please have a look? (adding the "cleanup" label for now)

**Comment by dylhunn** [comment]
If this must go in on its own, I will wait until tomorrow and do a second merge and sync.

**Pull request closed without merging**